File: src/main/java/com/nickcoblentz/montoya/settings/GenericExtensionSetting.java
package com.nickcoblentz.montoya.settings;

import burp.api.montoya.MontoyaApi;
import com.nickcoblentz.montoya.MontoyaLogger;

import java.util.List;

public abstract class GenericExtensionSetting<T> {

    private final T DefaultValue;
    private final String StorageKey;
    private final String Name;

    private final MontoyaApi Api;
    private final ExtensionSettingSaveLocation DefaultSaveLocation;
    public final MontoyaLogger Logger;
    private T CurrentValue;

    public GenericExtensionSetting(MontoyaApi api, String name, String storageKey, T defaultValue, ExtensionSettingSaveLocation defaultSaveLocation) {
        Logger = new MontoyaLogger(api,MontoyaLogger.DebugLogLevel);
        Api=api;
        Name=name;
        StorageKey=storageKey;
        DefaultValue=defaultValue;
        DefaultSaveLocation=defaultSaveLocation;
        load();
    }


    public T getCurrentValue() {
        if(CurrentValue !=null)
        {
            if(CurrentValue instanceof String sCurrentValue)
            {
                if(!sCurrentValue.isEmpty())
                    return CurrentValue;
            }
            else {
                return CurrentValue;
            }
        }

        return DefaultValue;
    }


    public void setCurrentValue(T currentValue) {
        if(currentValue!=null)
            CurrentValue = currentValue;
    }

    public abstract void setCurrentValue(String value);


    public abstract void saveToProject();


    public abstract void loadFromProject();/*
    {
        setCurrentValue(Api.persistence().extensionData().getString(StorageKey));
    }*/


    public abstract void saveToPreferences();


    public abstract void loadFromPreferences();


    public void save() {
        switch(DefaultSaveLocation) {
            case ExtensionSettingSaveLocation.PREFERENCE -> saveToPreferences();
            case ExtensionSettingSaveLocation.PROJECT -> saveToProject();
            default -> {}
        }
    }


    public void load() {
        switch(DefaultSaveLocation) {
            case ExtensionSettingSaveLocation.PREFERENCE -> loadFromPreferences();
            case ExtensionSettingSaveLocation.PROJECT -> loadFromProject();
            default -> {}
        }
    }


    public String getName() {
        return Name;
    }


    public String getStorageKey() {
        return StorageKey;
    }


    public T reset()
    {
        setCurrentValue(DefaultValue);
        return getCurrentValue();
    }

    public MontoyaApi getApi() {
        return Api;
    }

    public abstract T parseString(String value);

    public T getDefaultValue() {
        return DefaultValue;
    }
}

File: src/main/java/com/nickcoblentz/montoya/settings/GenericExtensionSetting.kt
package com.nickcoblentz.montoya.settings

import burp.api.montoya.MontoyaApi
import com.nickcoblentz.montoya.LogLevel
import com.nickcoblentz.montoya.MontoyaLogger

abstract class GenericExtensionSetting<T>(
    val api: MontoyaApi,
    val name: String,
    val storageKey: String,
    val defaultValue: T,
    private val defaultSaveLocation: ExtensionSettingSaveLocation
) {


    //Logger = new MontoyaLogger(api,MontoyaLogger.DebugLogLevel);
    val Logger: MontoyaLogger = MontoyaLogger(api, LogLevel.DEBUG)
    var currentValue = defaultValue
        get() {
            return when(field) {
                null -> defaultValue
                else -> field
            }
        }
        set(value) {
            if (value != null) field = value
        }

    init {
        load()
    }

    abstract fun setCurrentValue(value: String?)


    abstract fun saveToProject()


    abstract fun loadFromProject() /*
    {
        setCurrentValue(Api.persistence().extensionData().getString(StorageKey));
    }*/


    abstract fun saveToPreferences()


    abstract fun loadFromPreferences()


    fun save() {
        when (defaultSaveLocation) {
            ExtensionSettingSaveLocation.PREFERENCE -> saveToPreferences()
            ExtensionSettingSaveLocation.PROJECT -> saveToProject()
            else -> {}
        }
    }


    fun load() {
        when (defaultSaveLocation) {
            ExtensionSettingSaveLocation.PREFERENCE -> loadFromPreferences()
            ExtensionSettingSaveLocation.PROJECT -> loadFromProject()
            else -> {}
        }
    }


    fun reset(): T {
        currentValue=defaultValue
        return currentValue
    }

    abstract fun parseString(value: String): T
}